src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -246,100 +258,497 @@ private void InitializeFromCollection(IEnumerable<KeyValuePair<TKey, TValue>> co
+                                            (sync & (int)RecordStatus.HasValue3) != 0
+                                                        &&
+                                            comp.Equals(key, keys.Key3)
+                                        )
+                                    {
+                                        syncs[index] = sync;
+
+                                        return false;
+                                    }
+
+                                    keyst[index].Key2   = key;
+                                    vals[index].Value2  = value;
+                                    syncs[index]        = sync | (int)RecordStatus.HasValue2;
+                                }
+                                // add to cell 3
+                                else
+                                {
+                                    // check exist
+                                    if (
+                                            comp.Equals(key, keys.Key0)
+                                                        ||
+                                            comp.Equals(key, keys.Key1)
+                                                        ||
+                                            comp.Equals(key, keys.Key2)
+                                        )
+                                    {
+                                        syncs[index] = sync;
+
+                                        return false;
+                                    }
+
+                                    keyst[index].Key3   = key;
+                                    vals[index].Value3  = value;
+                                    syncs[index]        = sync | (int)RecordStatus.HasValue3;
+                                }
+
+                                IncrementCount(data);
+
+                                return true;
+                            }
+                            catch
+                            {
+                                syncs[index] = sync;
+
+                                throw;
+                            }
+                        }
+                    }
+                    // growing
+                    else
+                    {
+                        var keys = keyst[index];
+
+                        // check exist
+                        if (
+                                comp.Equals(key, keys.Key0)
+                                            ||
+                                comp.Equals(key, keys.Key1)
+                                            ||
+                                comp.Equals(key, keys.Key2)
+                                            ||
+                                comp.Equals(key, keys.Key3)
+                            )
+                        {
+                            return false;
+                        }
 
-            return TryRemoveInternal(key, out value, false, default);
+                        GrowTable(data, hash);
```

Comment:
Growing the table able after chain of 4 collisions works great for ideal distribution of hashcodes. It will lead to a much higher memory consumption if there is a small cluster of poorly distributed hashcodes.

Reply:
@jkotas 
I am rewriting the table now. Implements wait-free TryGetValue. And I will consider your wishes in the new version.